### PR TITLE
Dynamically search for the applied `@JvmLoggingDomain` annotations

### PR DESCRIPTION
**Pull request by yevhenii-nadtochii**
Addresses #39.

The issue caused by greedy analysis of presence of `@JvmLoggingDomain` on packages. The used `AnnotatedPackages` class scans packages during the initialization. And doesn't update set of loaded packages anymore. While packages are loaded dynamically as classes, on demand. Thus, it may end up with an outdated data about the annotated packages.

This PR introduces `PackageAnnotationLookup` class. It replaces `AnnotatedPackages` in `LoggingDomainClassValue`. Lookup is performed on demand, and the scanned packages are cached to speed up further requests. 

Each search includes scanning of loaded packages, and even force-loading of potentially annotated ones. It eliminates a possibility of false-negative return when a midway package is annotated, but not loaded.

### Class location

Though, `PackageAnnotationLookup` is developed in-place. Further, it should go to our [spine-reflect](https://github.com/SpineEventEngine/reflect) library and become public.

I will create an issue for that.

### CODE REVIEW

**Comment by yevhenii-nadtochii** [comment]
@alexander-yevsyukov @armiol Please take a look.

File: logging/src/jvmMain/kotlin/io/spine/reflect/PackageAnnotationLookup.kt [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+ *
+ * This lookup is similar to [AnnotatedPackages][io.spine.reflect.AnnotatedPackages].
+ *
+ * [AnnotatedPackages][io.spine.reflect.AnnotatedPackages] eagerly traverses
+ * all loaded packages during the instance creation. It looks for ones
+ * that are annotated with the given annotation type, remembers them,
+ * and then allows retrieving of an annotation for the asked package.
+ *
+ * But as more classes are loaded by the classloader, more new packages appear.
+ * As a result, data within the collection become insufficient. An instance
```
Comment (armiol) [inline logging/src/jvmMain/kotlin/io/spine/reflect/PackageAnnotationLookup.kt:45]: "... data within the collection become insufficient..."

This portion required rewriting. First of all, "data" is "it", not "they". Second of all, "insufficient" for what?

File: logging/src/jvmMain/kotlin/io/spine/reflect/PackageAnnotationLookup.kt [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+ * When the lookup inspects parental packages, they are also all cached.
+ * Since we have already fetched the loaded parental packages (or tried
+ * to force-load them), there's no reason not to cache them too.
+ *
+ * Traversing and caching of parental packages will continue even if we have
+ * already found the closest annotated parent. Checking two-three-five more
+ * packages is not costly when instances of [Package] are already at hand.
+ * Otherwise, it may cause many unnecessary repeated force-loadings.
+ */
+internal class PackageAnnotationLookup<T : Annotation>(
```
Comment (armiol) [inline logging/src/jvmMain/kotlin/io/spine/reflect/PackageAnnotationLookup.kt:64]: For the sake of law&order, please document `T` as a `@param`.

File: logging/src/jvmMain/kotlin/io/spine/reflect/PackageAnnotationLookup.kt [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+     * This map contains both directly annotated packages and ones
+     * that have any parent annotated with [T] (propagated annotation).
+     */
+    private val knownPackages = hashMapOf<PackageName, T?>()
+
+    init {
+        val annotations = wantedAnnotation.annotations
+        val isRepeatable = annotations.any { it.annotationClass == Repeatable::class }
+        require(isRepeatable.not()) {
+            "Lookup for repeatable annotations is not supported."
```
Comment (armiol) [inline logging/src/jvmMain/kotlin/io/spine/reflect/PackageAnnotationLookup.kt:102]: In such cases, let's include the annotation itself into error messages.

**Review by armiol — CHANGES_REQUESTED** [review]
@yevhenii-nadtochii please see my comments.

File: logging/src/jvmMain/kotlin/io/spine/reflect/JvmPackages.kt [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+     * io.spine.reflect
+     * io.spine
+     * io
+     * ```
+     *
+     * Please note, this method just operates upon the given package name.
+     * Its result is not guaranteed to match an existent hierarchy of packages.
+     */
+    fun expand(name: PackageName): List<PackageName> {
+        val buffer = StringBuilder(name.length)
```
[the remark this answers is not in the document]
Comment (yevhenii-nadtochii) [reply logging/src/jvmMain/kotlin/io/spine/reflect/JvmPackages.kt:108]: `split` would return components of a package: `[io, spine, reflect]`. 

What we need can be done with regex search. But it is simpler and faster to use a `StringBuilder`.

**Comment by yevhenii-nadtochii** [comment]
@armiol PTAL

**Review by armiol — APPROVED** [review]

**Pull request merged**